include/alpaka/pltf/PltfGenericSycl.hpp

```diff
@@ -18,89 +18,135 @@
 
 #    include <CL/sycl.hpp>
 
-#    include <iostream>
+#    include <cstddef>
+#    include <exception>
+#    include <functional>
+#    if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
+#        include <iostream>
+#    endif
+#    include <memory>
 #    include <sstream>
 #    include <stdexcept>
 #    include <vector>
 
 namespace alpaka::experimental
 {
     //! The SYCL device manager.
-    class PltfGenericSycl : public concepts::Implements<ConceptPltf, PltfGenericSycl>
+    template<typename TSelector>
+    class PltfGenericSycl : public concepts::Implements<ConceptPltf, PltfGenericSycl<TSelector>>
     {
     public:
         PltfGenericSycl() = delete;
+
+        [[nodiscard]] static auto syclPlatform() -> sycl::platform&
+        {
+            if(platform_ptr == nullptr)
+                platform_ptr = std::make_unique<sycl::platform>(TSelector{});
+
+            return *platform_ptr;
+        }
+
+        [[nodiscard]] static auto syclDevices() -> std::vector<sycl::device>&
+        {
+            if(devices_ptr == nullptr)
+                devices_ptr = std::make_unique<std::vector<sycl::device>>(syclPlatform().get_devices());
+
+            return *devices_ptr;
+        }
+
+        [[nodiscard]] static auto syclContext() -> sycl::context&
+        {
+            if(context_ptr == nullptr)
+            {
+                context_ptr = std::make_unique<sycl::context>(
+                    syclDevices(),
+                    [](sycl::exception_list exceptions)
+                    {
+                        auto ss_err = std::stringstream{};
+                        ss_err << "Caught asynchronous SYCL exception(s):\n";
+                        for(std::exception_ptr e : exceptions)
+                        {
+                            try
+                            {
+                                std::rethrow_exception(e);
+                            }
+                            catch(sycl::exception const& err)
+                            {
+                                ss_err << err.what() << " (" << err.code() << ")\n";
+                            }
+                        }
+                        throw std::runtime_error(ss_err.str());
+                    });
+            }
+
+            return *context_ptr;
+        }
+
+        // Use destructor attribute to make sure our static variables are destroyed in the correct order at shutdown.
+        [[gnu::destructor]] static auto reset() -> void
+        {
+            context_ptr.reset();
+            devices_ptr.reset();
+            platform_ptr.reset();
+        }
+
+    private:
+#    if BOOST_COMP_CLANG
+#        pragma clang diagnostic push
+#        pragma clang diagnostic ignored "-Wexit-time-destructors"
+#    endif
+        inline static std::unique_ptr<sycl::platform> platform_ptr{nullptr};
+        inline static std::unique_ptr<std::vector<sycl::device>> devices_ptr{nullptr};
+        inline static std::unique_ptr<sycl::context> context_ptr{nullptr};
```

Comment:
Why not `std::optional` instead of `std::unique_ptr`?

Reply:
Old habits die hard...